Storage/Storage/CoreData/CoreDataManager.swift

```diff
@@ -152,14 +160,16 @@ public final class CoreDataManager: StorageManagerType {
     /// This method effectively destroys all of the stored data, and generates a blank Persistent Store from scratch.
     ///
     public func reset() {
-        let viewContext = persistentContainer.viewContext
+        syncQueue.sync {
+            let viewContext = persistentContainer.viewContext
 
-        viewContext.performAndWait {
-            viewContext.reset()
-            self.deleteAllStoredObjects()
+            viewContext.performAndWait {
+                viewContext.reset()
+                self.deleteAllStoredObjects()
 
-            DDLogVerbose("💣 [CoreDataManager] Stack Destroyed!")
-            NotificationCenter.default.post(name: .StorageManagerDidResetStorage, object: self)
+                DDLogVerbose("💣 [CoreDataManager] Stack Destroyed!")
+                NotificationCenter.default.post(name: .StorageManagerDidResetStorage, object: self)
+            }
         }
     }
```

Comment:
I wonder if there is a potential dead lock here.

Imagine `reset` is called on the main thread. The main thread now waits on the `syncQueue` to finish running. The `syncQueue` waits `viewContext` to complete. If the `viewContext` is bound to the main context, it waits for the main thread, which is blocked.

The above scenario is pretty easy to set up in unit tests. I'd suggesting adding an unit test to see if there is a dead-lock here.